### PR TITLE
[Xedra Evolved] Add time magic class / scenario

### PR DESCRIPTION
**Pull request by b3brodie**
#### Summary
Mods "[Xedra Evolved] Add time magic class / scenario"

#### Purpose of change
Add a new time magic class and scenario that offers new types of abilities / gameplay and forces you to use those powers to survive the negative effects that the scenario will start you with

#### Describe the solution
Add a scenario that starts the character off further into the cataclysm with the new class with the potential for more power but starting off with some significant downsides.

Start off by designing powers since later additions will make it a little harder to test them

Additions:
Spells:
- Time Freeze: Freeze the target in time, making them invulnerable to damage and incapable of acting
- Unfreeze Time: Unfreeze any target from time and / or unfreeze any individual fields of frozen time
- Entropic Burst: Unfreeze any targets and fields of frozen time in an aoe and apply a dot to targets within
- Time Bubble: Freeze targets and the air itself in an aoe, generating a large field of normally impassable frozen time invulnerable to any attempts to change the contents.
- Revert Wound: Reverts the last wound you've taken, removing the damage, pain, and potential bleeding
- Time Loop: Continuously reset yourself to when you've casted the spell.  Loop is broken by death.
- Revert Location: Translocate Ability + Translocator Construction
- Cement Time: Make a time effect permanent.  (Frozen tile, frozen monster, etc).
- Stable Time Loop (T): Save your current character state to allow you to reset to it by casting the spell again.  This is a toggled power that costs maximum mana.
- Chronal Acceleration (T): Improves character speed.  This is a toggled power that costs maximum mana.
- Destabilizing Strikes (T): Makes melee attacks apply a dot.  This is a toggled power that costs maximum mana.
- Stabilize Timeline (T): Makes the user immune to time freezes and allows traversal through frozen time fields.  Does not allow modifying things within the time field.  This is a toggled power that costs maximum mana.
- Time Stop: Stops time for everything other that the user for a certain amount of time
- Reverse Entropy (T): Slowly reverts equipment to peak shape, repairing damage and recharging energy.  This is a toggled power that costs maximum mana.
- Rewrite Wound Causality (M): Meditate to spend mana to heal the character of their wounds
- Rewrite Equipment Causality (M): Meditate to spend mana to repair equipment of durability damage.

Starting Problem Traits:
- Time Looped Knife Wound: Every 24 hours, the character suffers a knife wound to their torso causing damage, pain, and bleeding.
- Slowed Timestream: Reduce speed by 10%
- Unstable Powers: When using toggled powers, every 30 minutes - 2 hours there is a chance they all fail, increased by the amount currently toggled.
- Not a trait, but the character starts during winter with higher evolved enemies and no winter clothes.

Spell Learning:
- Spell Learning menu similar to the Bombastic Perks menu.  Spend earned insight to learn new powers.
- Also allows removing negative chronomancer traits by spending insight.
- Killing time related enemies gives chronomantic insight

Infrastructure:
- Adds CANNOT_GAIN_EFFECTS flag that makes things not able to gain effects.
- Allows checking and modifying item degradation with eocs
- Adds indestructible flag as an option for field_types.  Prevents destroying fields by bashing, shooting into them, or reach attacking through them.

#### Testing
Created a character and checked that the starting modifiers occurred correctly
Used the insight tree to gain powers correctly
Killed enemies and gained insight correctly
A good amount of sanity checking with time loops to minimize buggy edge cases
Checked that the downside mutations functioned correctly

### CODE REVIEW

**Review by andrei8l — COMMENTED** [review]
`parse_output_tags` is madness and needs to go - we can already handle that use case with var_val. 

This PR does a lot of different things and is hard to review

File: src/creature.cpp [hunk 1]
```diff
@@ -1681,7 +1682,7 @@ void Creature::add_effect( const effect_source &source, const efftype_id &eff_id
                            const time_duration &dur, bodypart_id bp, bool permanent, int intensity, bool force, bool deferred )
 {
     // Check our innate immunity
-    if( !force && is_immune_effect( eff_id ) ) {
+    if( !force && is_immune_effect( eff_id ) || has_flag( json_flag_CANNOT_GAIN_EFFECTS ) ) {
```
[the remark this answers is not in the document]
Comment (b3brodie) [reply src/creature.cpp]: Added the parenthesis around the and condition in this case.

File: src/creature.h [hunk 2]
```diff
@@ -765,6 +765,9 @@ class Creature : public viewer
         virtual bool has_flag( const mon_flag_id & ) const {
             return false;
         }
+        virtual bool has_flag( const flag_id & ) const {
+            return false;
+        }
```
[the remark this answers is not in the document]
Comment (b3brodie) [reply src/creature.h:770]: Thanks.  I was having trouble figuring out the syntax for this one.

**Comment by b3brodie** [comment]
> `parse_output_tags` is madness and needs to go - we can already handle that use case with var_val.
> 
> This PR does a lot of different things and is hard to review

Sorry about that.  I was considering breaking some of it up into separate PRs but I wasn't sure if it would be better or not.  If it would help I can definitely split the monsters and related changes into a separate PR, along with maybe a couple of other things

**Pull request merged**